smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/traits/EndpointTestsTraitValidator.java

```diff
@@ -47,42 +71,102 @@ public List<ValidationEvent> validate(Model model) {
+                            testCase,
+                            testOperationInput,
+                            events);
+
+                    StructureShape inputShape = model.expectShape(
+                            operationNameMap.get(operationName).getInputShape(),
+                            StructureShape.class);
+                    validateOperationInput(model, serviceShape, inputShape, testCase, testOperationInput, events);
                 }
             }
         }
 
         return events;
     }
 
-    private List<ValidationEvent> validateOperationInput(
+    private Map<Parameter, Node> getMismatchedBuiltInParams(
+            List<Parameter> builtInParamsWithDefaults,
+            EndpointTestCase testCase
+    ) {
+        Map<Parameter, Node> mismatchedBuiltInParams = new HashMap<>();
+        for (Parameter parameter : builtInParamsWithDefaults) {
+            String parameterName = parameter.getName().toString();
+            Node defaultValue = parameter.getDefault().get().toNode();
+
+            // Consider a parameter mismatched if the built-in's default and its
+            // value in this test case aren't the same.
+            if (testCase.getParams().containsMember(parameterName)
+                    && !testCase.getParams().expectMember(parameterName).equals(defaultValue)) {
+                mismatchedBuiltInParams.put(parameter, testCase.getParams().expectMember(parameterName));
+            }
+        }
+        return mismatchedBuiltInParams;
+    }
+
+    private void validateBuiltInValuesAreValid(
+            ServiceShape serviceShape,
+            Map<Parameter, Node> mismatchedBuiltInParams,
+            EndpointTestOperationInput testOperationInput,
+            List<ValidationEvent> events
+    ) {
+        for (Map.Entry<Parameter, Node> mismatchedBuiltInParam : mismatchedBuiltInParams.entrySet()) {
+            String builtInName = mismatchedBuiltInParam.getKey().getBuiltIn().get();
+            // Emit if either the built-in with a mismatched value isn't
+            // specified or the value set for it doesn't match.
+            if (!testOperationInput.getBuiltInParams().containsMember(builtInName)
+                    || !testOperationInput.getBuiltInParams()
+                            .expectMember(builtInName)
+                            .equals(mismatchedBuiltInParam.getValue())) {
+                events.add(error(serviceShape,
+                        testOperationInput,
+                        String.format("Test case does not supply the `%s` value for the `%s` parameter's "
+                                + "`%s` built-in.",
+                                Node.printJson(mismatchedBuiltInParam.getValue()),
+                                mismatchedBuiltInParam.getKey().getNameString(),
+                                mismatchedBuiltInParam.getKey().getBuiltIn().get())));
+            }
+        }
+    }
+
+    private void validateNoDefaultBuiltInParamsHaveValues(
```

Comment:
Sorry, another slight naming nit - I read this name as validating that built ins with defaults do not have values.  Maybe something like `validateBuiltInsWithoutDefaultsHaveValues`.  I think the logic below is correct, I just had to read it a few times to figure out what the method was supposed to be validating.